doc/man/cgconfig.conf.5
.TH CGCONFIG.CONF 5
.\"***********************************
.SH NAME
cgconfig.conf \- libcgroup configuration file
.\"***********************************
.SH DESCRIPTION
.B "cgconfig.conf"
is a configuration file used by
.B libcgroup
to define control groups, their parameters and their mount points.
The file consists of
.I mount
,
.I group
,
.I systemd
, and
.I default
sections. These sections can be in arbitrary order and all of them are
optional. Any line starting with '#' is considered a comment line and
is ignored.
.LP
.I mount
section has this form:
.RS
.nf
.ft B
.sp
mount {
.RS
.ft B
<controller> = <path>;
.I "..."
.RE
.ft B
}
.ft R
.fi
.RE

.TP
.B controller
Name of the kernel subsystem. The list of subsystems supported by the kernel
can be found in
.I /proc/cgroups
file. Named hierarchy can be specified as controller
\fB"name=<somename>"\fR. Do not forget to use double quotes around
this controller name (see examples below). Apart from named hierarchy,
additional mount options may be specified by putting the controller and
the options in quotes. Options supported are \fB nosuid, noexec\fR and\fB nodev\fR.

.B Libcgroup
merges all subsystems mounted to the same directory (see
Example 1) and the directory is mounted only once.

.TP
.B path
The directory path where the group hierarchy associated to a given
controller shall be mounted. The directory is created
automatically on cgconfig service startup if it does not exist and
is deleted on service shutdown.
.LP

If no
.I mount
section is specified, no controllers are mounted.

.I group
section has this form:
.RS
.nf
.ft B
.sp
group <name> {
.RS
.ft B
[permissions]
<controller> {
.RS
.ft B
<param name> = <param value>;
.I "..."
.RE
.ft B
}
.I "..."
.RE
.ft B
}
.ft R
.fi
.RE

.TP
.B name
Name of the control group. It can contain only characters, which are
allowed for directory names.
The groups form a tree, i.e. a control group can contain zero or more
subgroups. Subgroups can be specified using '/' delimiter.

The root control group is always created automatically in all hierarchies
and it is the base of the group hierarchy. It can be explicitly specified in
.B cgconfig.conf
by using '.' as group name. This can be used e.g. to set its permissions,
as shown in Example 6.

When the parent control group of a subgroup is not specified
it is created automatically.

.TP
.B permissions
Permissions of the given control group on mounted filesystem.
.I root
has always permission to do anything with the control group.
Permissions have the following syntax:
.RS 17
.ft B
.nf
perm {
.RS
.ft B
task {
.RS
.ft B
uid = <task user>;
gid = <task group>;
fperm = <file permissions>
.RE
}
admin {
.RS
uid = <admin name>;
gid = <admin group>;
dperm = <directory permissions>
fperm = <file permissions>
.RE
}
.RE
}
.fi
.RE
.ft R

.RS
.TP 17
.B "task user/group"
Name of the user and the group, which own the
.I tasks
file of the control group. Given fperm then specify the file permissions.
Please note that the given value is not used as was specified. Instead,
current file owner permissions are used as a "umask" for group and others
permissions. For example if fperm = 777 then both group and others will get
the same permissions as the file owner.
.TP 17
.B "admin user/group"
Name of the user and the group which own the rest of control group's
files. Given fperm and dperm control file and directory permissions.
Again, the given value is masked by the file/directory owner permissions.
.LP
Permissions are only apply to the enclosing control group and are not
inherited by subgroups. If there is no
.B perm
section in the control group definition,
.I root:root
is the owner of all files and default file permissions are preserved if
fperm resp. dperm are not specified.
.RE
.TP
.B controller
Name of the kernel subsystem.
The section can be
empty, default kernel parameters will be used in this case. By
specifying
.B controller
the control group and all its parents are controlled by the specific
subsystem. One control group can be controlled by multiple subsystems,
even if the subsystems are mounted on different directories. Each
control group must be controlled by at least one subsystem, so that
.B libcgroup
knows in which hierarchies the control group should be created.

The parameters of the given controller can be modified in the following
section enclosed in brackets.
.RS
.TP
.B param name
Name of the file to set. Each controller can have zero or more
parameters.
.TP
.B param value
Value which should be written to the file when the control group is
created. If it is enclosed in double quotes `"', it can contain spaces
and other special characters.
.RE

If no
.I group
section is specified, no groups are created.

.I systemd
section has this form:
.RS
.nf
.ft B
.sp
systemd {
.RS
.ft B
slice = <slice name>;
scope = <scope name>;
[setdefault = true;]
[pid = <pid>;]
.RE
.ft B
}
.ft R
.fi
.RE

Multiple systemd sections are supported.

.TP
.B slice name
Name of the systemd slice where the scope will reside. If the slice does not
exist, then systemd will create it when the scope is created. The slice name
must be of the form *.slice, e.g. libcgroup.slice.

.TP
.B scope name
Name of the systemd scope that will be created. If the scope already exists,
then cgconfigparser will not overwrite the existing scope, and the appropriate
error will be returned to the user. The scope name must be of the form
*.scope, e.g. database.scope.

.TP
.B setdefault
If set, then libcgroup will use the scope and slice in this section as the
default cgroup path. If setdefault is true in multiple systemd sections,
libcgroup uses the last one parsed as its default and previous defaults are
discarded. This default value is used by most libcgroup tools - cgget, etc. -
and can be overridden if necessary. See Example 9 below.

.TP
.B pid
Process ID to be placed in the scope. Systemd will remove any scope that
doesn't have at least one process in it. If a PID is provided via the
\fBpid\fR tag, then libcgroup will place that PID in the scope. Otherwise,
libcgroup will create a process that sleeps infinitely and place it in the
scope.
.RE

.I default
section has this form:
.RS
.nf
.ft B
.sp
default {
.RS
.ft B
perm {
.RS
.ft B
task {
.RS
.ft B
uid = <task user>;
gid = <task group>;
fperm = <file permissions>
.RE
}
admin {
.RS
uid = <admin name>;
gid = <admin group>;
dperm = <directory permissions>
fperm = <file permissions>
.RE
}
.RE
}
.RE
}
.ft R
.fi
.RE

Content of the
.B perm
section has the same form as in
.I group
section. The permissions defined here specify owner and permissions of
groups and files of all groups, which do not have explicitly specified
their permissions in their
.I group
section.

.I template
section has the same structure as
.B group
section. Template name uses the same templates string as
.B cgrules.conf
destination tag (see (\fBcgrules.conf\fR (5)).
Template definition is used as a control group definition for rules in
\fBcgrules.conf\fR (5) with the same destination name.
Templates does not use
.B default
section settings.

.I /etc/cgconfig.d/
directory can be used for additional configuration files. cgrulesengd searches this directory for additional templates.

.\"********************************************"
.SH EXAMPLES
.LP
.SS Example 1
.LP
The configuration file:
.LP
.RS
.nf
mount {
.RS
cpu = /mnt/cgroups/cpu;
cpuacct = /mnt/cgroups/cpu;
.RE
}
.fi
.RE

creates the hierarchy controlled by two subsystems with no groups
inside. It corresponds to the following operations:
.LP
.RS
.nf
mkdir /mnt/cgroups/cpu
mount -t cgroup -o cpu,cpuacct cpu /mnt/cgroups/cpu
.fi
.RE

.SS Example 2
.LP
The configuration file:
.LP
.RS
.nf
mount {
.RS
cpu = /mnt/cgroups/cpu;
"name=scheduler" = /mnt/cgroups/cpu;
"name=noctrl" = /mnt/cgroups/noctrl;
.RE
}

group daemons {
.RS
cpu {
.RS
cpu.shares = "1000";
.RE
}
.RE
}
group test {
.RS
"name=noctrl" {
}
.RE
}
.RE
.fi
creates two hierarchies. One hierarchy named \fBscheduler\fR controlled by cpu
subsystem, with group \fBdaemons\fR inside. Second hierarchy is named
\fBnoctrl\fR without any controller, with group \fBtest\fR. It corresponds to
following operations:
.LP
.RS
.nf
mkdir /mnt/cgroups/cpu
mount -t cgroup -o cpu,name=scheduler cpu /mnt/cgroups/cpu
mount -t cgroup -o none,name=noctrl none /mnt/cgroups/noctrl

mkdir /mnt/cgroups/cpu/daemons
echo 1000 > /mnt/cgroups/cpu/daemons/www/cpu.shares

mkdir /mnt/cgroups/noctrl/tests
.fi
.RE

The
.I daemons
group is created automatically when its first subgroup is
created. All its parameters have the default value and only root can
access group's files.
.LP
Since both
.I cpuacct
and
.I cpu
subsystems are mounted to the same directory, all
groups are implicitly controlled also by
.I cpuacct
subsystem, even if there is no
.I cpuacct
section in any of the groups.
.RE

.SS Example 3
.LP
The configuration file:
.LP
.RS
.nf
mount {
.RS
cpu = /mnt/cgroups/cpu;
cpuacct = /mnt/cgroups/cpu;
.RE
}

group daemons/www {
.RS
perm {
.RS
task {
.RS
uid = root;
gid = webmaster;
fperm = 770;
.RE
}
admin {
.RS
uid = root;
gid = root;
dperm = 775;
fperm = 744;
.RE
}
.RE
}
cpu {
.RS
cpu.shares = "1000";
.RE
}
.RE
}

group daemons/ftp {
.RS
perm {
.RS
task {
.RS
uid = root;
gid = ftpmaster;
fperm = 774;
.RE
}
admin {
.RS
uid = root;
gid = root;
dperm = 755;
fperm = 700;
.RE
}
.RE
}
cpu {
.RS
cpu.shares = "500";
.RE
}
.RE
}
.RE
.fi
creates the hierarchy controlled by two subsystems with one group and
two subgroups inside, setting one parameter.
It corresponds to the following operations (except for file permissions
which are little bit trickier to emulate via chmod):

.LP
.RS
.nf
mkdir /mnt/cgroups/cpu
mount -t cgroup -o cpu,cpuacct cpu /mnt/cgroups/cpu

mkdir /mnt/cgroups/cpu/daemons

mkdir /mnt/cgroups/cpu/daemons/www
chown root:root /mnt/cgroups/cpu/daemons/www/*
chown root:webmaster /mnt/cgroups/cpu/daemons/www/tasks
echo 1000 > /mnt/cgroups/cpu/daemons/www/cpu.shares

 # + chmod the files so the result looks like:
 # ls -la /mnt/cgroups/cpu/daemons/www/
 # admin.dperm = 755:
 # drwxr-xr-x. 2 root webmaster 0 Jun 16 11:51 .
 #
 # admin.fperm = 744:
 # --w-------. 1 root webmaster 0 Jun 16 11:51 cgroup.event_control
 # -r--r--r--. 1 root webmaster 0 Jun 16 11:51 cgroup.procs
 # -r--r--r--. 1 root webmaster 0 Jun 16 11:51 cpuacct.stat
 # -rw-r--r--. 1 root webmaster 0 Jun 16 11:51 cpuacct.usage
 # -r--r--r--. 1 root webmaster 0 Jun 16 11:51 cpuacct.usage_percpu
 # -rw-r--r--. 1 root webmaster 0 Jun 16 11:51 cpu.rt_period_us
 # -rw-r--r--. 1 root webmaster 0 Jun 16 11:51 cpu.rt_runtime_us
 # -rw-r--r--. 1 root webmaster 0 Jun 16 11:51 cpu.shares
 # -rw-r--r--. 1 root webmaster 0 Jun 16 11:51 notify_on_release
 #
 # tasks.fperm = 770
 # -rw-rw----. 1 root webmaster 0 Jun 16 11:51 tasks


mkdir /mnt/cgroups/cpu/daemons/ftp
chown root:root /mnt/cgroups/cpu/daemons/ftp/*
chown root:ftpmaster /mnt/cgroups/cpu/daemons/ftp/tasks
echo 500 > /mnt/cgroups/cpu/daemons/ftp/cpu.shares

 # + chmod the files so the result looks like:
 # ls -la /mnt/cgroups/cpu/daemons/ftp/
 # admin.dperm = 755:
 # drwxr-xr-x. 2 root ftpmaster 0 Jun 16 11:51 .
 #
 # admin.fperm = 700:
 # --w-------. 1 root ftpmaster 0 Jun 16 11:51 cgroup.event_control
 # -r--------. 1 root ftpmaster 0 Jun 16 11:51 cgroup.procs
 # -r--------. 1 root ftpmaster 0 Jun 16 11:51 cpuacct.stat
 # -rw-------. 1 root ftpmaster 0 Jun 16 11:51 cpuacct.usage
 # -r--------. 1 root ftpmaster 0 Jun 16 11:51 cpuacct.usage_percpu
 # -rw-------. 1 root ftpmaster 0 Jun 16 11:51 cpu.rt_period_us
 # -rw-------. 1 root ftpmaster 0 Jun 16 11:51 cpu.rt_runtime_us
 # -rw-------. 1 root ftpmaster 0 Jun 16 11:51 cpu.shares
 # -rw-------. 1 root ftpmaster 0 Jun 16 11:51 notify_on_release
 #
 # tasks.fperm = 774:
 # -rw-rw-r--. 1 root ftpmaster 0 Jun 16 11:51 tasks

.fi
.RE

The
.I daemons
group is created automatically when its first subgroup is
created. All its parameters have the default value and only root can
access the group's files.
.LP
Since both
.I cpuacct
and
.I cpu
subsystems are mounted to the same directory, all
groups are implicitly also controlled by the
.I cpuacct
subsystem, even if there is no
.I cpuacct
section in any of the groups.
.RE

.SS Example 4
.LP
The configuration file:

.LP
.RS
.nf
mount {
.RS
cpu = /mnt/cgroups/cpu;
cpuacct = /mnt/cgroups/cpuacct;
.RE
}

group daemons {
.RS
cpuacct{
}
cpu {
}
.RE
}
.fi
.RE
creates two hierarchies and one common group in both of them.
It corresponds to the following operations:
.LP
.RS
.nf
mkdir /mnt/cgroups/cpu
mkdir /mnt/cgroups/cpuacct
mount -t cgroup -o cpu cpu /mnt/cgroups/cpu
mount -t cgroup -o cpuacct cpuacct /mnt/cgroups/cpuacct

mkdir /mnt/cgroups/cpu/daemons
mkdir /mnt/cgroups/cpuacct/daemons
.fi
.RE

In fact there are two groups created. One in the
.I cpuacct
hierarchy, the second in the
.I cpu
hierarchy. These two groups have nothing in common and can
contain different subgroups and different tasks.

.SS Example 5
.LP

The configuration file:

.LP
.RS
.nf
mount {
.RS
cpu = /mnt/cgroups/cpu;
cpuacct = /mnt/cgroups/cpuacct;
.RE
}

group daemons {
.RS
cpuacct{
}
.RE
}

group daemons/www {
.RS
cpu {
.RS
cpu.shares = "1000";
.RE
}
.RE
}

group daemons/ftp {
.RS
cpu {
.RS
cpu.shares = "500";
.RE
}
.RE
}
.fi
.RE
creates two hierarchies with few groups inside. One of the groups
is created in both hierarchies.

It corresponds to the following operations:
.LP
.RS
.nf
mkdir /mnt/cgroups/cpu
mkdir /mnt/cgroups/cpuacct
mount -t cgroup -o cpu cpu /mnt/cgroups/cpu
mount -t cgroup -o cpuacct cpuacct /mnt/cgroups/cpuacct

mkdir /mnt/cgroups/cpuacct/daemons
mkdir /mnt/cgroups/cpu/daemons
mkdir /mnt/cgroups/cpu/daemons/www
echo 1000 > /mnt/cgroups/cpu/daemons/www/cpu.shares
mkdir /mnt/cgroups/cpu/daemons/ftp
echo 500 > /mnt/cgroups/cpu/daemons/ftp/cpu.shares
.fi
.RE
Group
.I daemons
is created in both hierarchies. In the
.I cpuacct
hierarchy the group is explicitly mentioned in the configuration
file. In the
.I cpu
hierarchy the group is created implicitly when
.I www
is created there. These two groups have nothing in common, for example
they do not share processes and subgroups. Groups
.I www
and
.I ftp
are created only in the
.I cpu
hierarchy and are not controlled by the
.I cpuacct
subsystem.

.SS Example 6
.LP
The configuration file:
.LP
.RS
.nf
mount {
.RS
cpu = /mnt/cgroups/cpu;
cpuacct = /mnt/cgroups/cpu;
.RE
}

group . {
.RS
perm {
.RS
task {
.RS
uid = root;
gid = operator;
.RE
}
admin {
.RS
uid = root;
gid = operator;
.RE
}
.RE
}
cpu {
}
.RE
}

group daemons {
.RS
perm {
.RS
task {
.RS
uid = root;
gid = daemonmaster;
.RE
}
admin {
.RS
uid = root;
gid = operator;
.RE
}
.RE
}
cpu {
}
.RE
}
.RE
.fi
creates the hierarchy controlled by two subsystems with one group having some
special permissions.
It corresponds to the following operations:
.LP
.RS
.nf
mkdir /mnt/cgroups/cpu
mount -t cgroup -o cpu,cpuacct cpu /mnt/cgroups/cpu

chown root:operator /mnt/cgroups/cpu/*
chown root:operator /mnt/cgroups/cpu/tasks

mkdir /mnt/cgroups/cpu/daemons
chown root:operator /mnt/cgroups/cpu/daemons/*
chown root:daemonmaster /mnt/cgroups/cpu/daemons/tasks
.fi
.RE

Users which are members of the
.I operator
group are allowed to administer the control groups, i.e. create new control
groups and move processes between these groups without having root
privileges.

Members of the
.I daemonmaster
group can move processes to the
.I daemons
control group, but they can not move the process out of the group. Only the
.I operator
or root can do that.

.SS Example 7
.LP
The configuration file:

.LP
.RS
.nf
mount {
.RS
cpu = /mnt/cgroups/cpu;
cpuacct = /mnt/cgroups/cpuacct;
.RE
}

group students {
.RS
cpuacct{
}
cpu {
}
.RE
}

template students/%u {
.RS
cpuacct{
}
cpu {
}
.RE
}

mkdir /mnt/cgroups/cpu/daemons
mkdir /mnt/cgroups/cpuacct/daemons
.fi
.RE

The situation is the similar as in Example 4. The only difference is template,
which is used if some rule uses "/students/%u" as a destination.

.SS Example 8
.LP
The configuration file:

.LP
.RS
.nf
mount {
.RS
"cpu,nodev,nosuid,noexec" = /mnt/cgroups/cpu;
.RE
}

.fi
.RE

This is the same as
mount -t cgroup cgroup -o nodev,nosuid,noexec,cpu /mnt/cgroups/cpu
It mounts the cpu controller with MS_NODEV, MS_NOSUID and MS_NOEXEC
options passed.

.SS Example 9
.LP
The configuration file:

.LP
.RS
.nf
systemd {
.RS
slice = myproduct.slice
scope = database1.scope
setdefault = true
.RE
}

.fi
.RE

creates the cgroups myproduct.slice/database1.scope. libcgroup instructs
systemd that the database1.scope is delegated, and the user is allowed to
fully modify this cgroup, its processes, and its settings/values. Since no
PID was provided in the configuration file, libcgroup will place a dummy
process that sleeps infinitely in the cgroup. systemd will automatically
delete database1.scope when there are no processes in it or its children.

Since setdefault was provided, the libcgroup tools will utilize the slice and
scope provided in this configuration when operating on the cgroup hierarchy.
For example, the following command:

.LP
.RS
.nf
$ cgcreate -g cpu:bgprocesses
.fi
.RE

will create a cgroup at /sys/fs/cgroup/myproduct.slice/database1.scope/bgprocesses
(assuming the cgroup hierarchy is mounted at /sys/fs/cgroup). This cgroup
can be interacted with via cgget/cgset as usual:

.LP
.RS
.nf
$ cgset -r cpu.weight=42 bgprocesses
$ cgget -r cpu.weight bgprocesses
bgprocesses:
cpu.weight 42
.fi
.RE

If a user wants to operate outside of the default systemd hierachy that was
specified in the configuration file above, they can do so as follows:

.LP
.RS
.nf
$ cgcreate -b -g cpu:outsidesystemd
.fi
.RE

This command will create a cgroup at /sys/fs/cgroup/outsidesystemd. Note that
this violates the kernel's and systemd's single writer rule, and systemd
reserves the right to modify this cgroup and its processes. \fBProceed at your
own risk.\fR This cgroup can be interacted with via cgget/cgset, note the
\fB-b\fR flag:

.LP
.RS
.nf
$ cgset -b -r cpu.weight=123 outsidesystemd
$ cgget -b -r cpu.weight outsidesystemd
outsidesystemd:
cpu.weight 123
.fi
.RE


.SH RECOMMENDATIONS
.SS Keep hierarchies separated
Having multiple hierarchies is perfectly valid and can be useful
in various scenarios. To keeps things clean, do not
create one group in multiple hierarchies. Examples 4 and 5 show
how unreadable and confusing it can be, especially when reading
somebody elses configuration file.

.SS Explicit is better than implicit
.B libcgroup
can implicitly create groups which are needed for the creation of
configured subgroups. This may be useful and save some typing in
simple scenarios. When it comes to multiple hierarchies, it's
better to explicitly specify all groups and all controllers
related to them.

.SH FILES
.TP
.B /etc/cgconfig.conf
default libcgroup configuration file
.TP
.B /etc/cgconfig.d/
default libcgroup configuration files directory

.SH SEE ALSO
cgconfigparser (8)

.SH BUGS
Parameter values must be single strings without spaces.
Parsing of quoted strings is not implemented.

.SH



src/tools/cgconfig.c

/*
 * Copyright IBM Corporation. 2007
 *
 * Authors:	Dhaval Giani <dhaval@linux.vnet.ibm.com>
 * 		Balbir Singh <balbir@linux.vnet.ibm.com>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2.1 of the GNU Lesser General Public License
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Code initiated and designed by Dhaval Giani. All faults are most likely
 * his mistake.
 */

#include <libcgroup.h>
#include <libcgroup-internal.h>

/* For basename() */
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
#include <string.h>
#include <libgen.h>

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <getopt.h>
#include "tools-common.h"

static struct cgroup_string_list cfg_files;

static void usage(int status, char *progname)
{
	if (status != 0) {
		fprintf(stderr, "Wrong input parameters, "\
			"try %s -h' for more information.\n", progname);
		return;
	}
	printf("Usage: %s [-h] [-f mode] [-d mode] [-s mode] "\
		"[-t <tuid>:<tgid>] [-a <agid>:<auid>] [-l FILE] "\
		"[-L DIR] ...\n", progname);
	printf("Parse and load the specified cgroups configuration file\n");
	printf("  -a <tuid>:<tgid>		Default owner of groups files "\
		"and directories\n");
	printf("  -d, --dperm=mode		Default group directory "\
		"permissions\n");
	printf("  -f, --fperm=mode		Default group file "\
		"permissions\n");
	printf("  -h, --help			Display this help\n");
	printf("  -l, --load=FILE		Parse and load the cgroups "\
		"configuration file\n");
	printf("  -L, --load-directory=DIR	Parse and load the cgroups "\
		"configuration files from a directory\n");
	printf("  -s, --tperm=mode		Default tasks file "\
		"permissions\n");
	printf("  -t <tuid>:<tgid>		Default owner of the tasks "\
		"file\n");
}

int main(int argc, char *argv[])
{
	int c, i;
	int ret, error = 0;
	static struct option options[] = {
		{"help", 0, 0, 'h'},
		{"load", 1, 0, 'l'},
		{"load-directory", 1, 0, 'L'},
		{"task", required_argument, NULL, 't'},
		{"admin", required_argument, NULL, 'a'},
		{"dperm", required_argument, NULL, 'd'},
		{"fperm", required_argument, NULL, 'f' },
		{"tperm", required_argument, NULL, 's' },
		{0, 0, 0, 0}
	};
	uid_t tuid = NO_UID_GID, auid = NO_UID_GID;
	gid_t tgid = NO_UID_GID, agid = NO_UID_GID;
	mode_t dir_mode = NO_PERMS;
	mode_t file_mode = NO_PERMS;
	mode_t tasks_mode = NO_PERMS;
	int dirm_change = 0;
	int filem_change = 0;
	struct cgroup *default_group = NULL;

	cgroup_set_default_logger(-1);

	if (argc < 2) {
		usage(1, argv[0]);
		return -1;
	}

	error = cgroup_string_list_init(&cfg_files, argc/2);
	if (error)
		goto err;

	while ((c = getopt_long(argc, argv, "hl:L:t:a:d:f:s:", options,
			NULL)) > 0) {
		switch (c) {
		case 'h':
			usage(0, argv[0]);
			error = 0;
			goto err;
		case 'l':
			error = cgroup_string_list_add_item(&cfg_files, optarg);
			if (error) {
				fprintf(stderr, "%s: cannot add file to list,"\
						" out of memory?\n", argv[0]);
				goto err;
			}
			break;
		case 'L':
			cgroup_string_list_add_directory(&cfg_files, optarg,
					argv[0]);
			break;
		case 'a':
			/* set admin uid/gid */
			error = parse_uid_gid(optarg, &auid, &agid, argv[0]);
			if (error)
				goto err;
			break;
		case 't':
			/* set task uid/gid */
			error = parse_uid_gid(optarg, &tuid, &tgid, argv[0]);
			if (error)
				goto err;
			break;
		case 'd':
			dirm_change = 1;
			error = parse_mode(optarg, &dir_mode, argv[0]);
			if (error)
				goto err;
			break;
		case 'f':
			filem_change = 1;
			error = parse_mode(optarg, &file_mode, argv[0]);
			if (error)
				goto err;
			break;
		case 's':
			filem_change = 1;
			error = parse_mode(optarg, &tasks_mode, argv[0]);
			if (error)
				goto err;
			break;
		default:
			usage(1, argv[0]);
			error = -1;
			goto err;
		}
	}

	if (argv[optind]) {
		usage(1, argv[0]);
		error = -1;
		goto err;
	}

	/* set default permissions */
	default_group = cgroup_new_cgroup("default");
	if (!default_group) {
		error = -1;
		fprintf(stderr, "%s: cannot create default cgroup\n", argv[0]);
		goto err;
	}

	error = cgroup_set_uid_gid(default_group, tuid, tgid, auid, agid);
	if (error) {
		fprintf(stderr, "%s: cannot set default UID and GID: %s\n",
				argv[0], cgroup_strerror(error));
		goto free_cgroup;
	}

	if (dirm_change | filem_change) {
		cgroup_set_permissions(default_group, dir_mode, file_mode,
				tasks_mode);
	}

	error = cgroup_config_set_default(default_group);
	if (error) {
		fprintf(stderr, "%s: cannot set config parser defaults: %s\n",
				argv[0], cgroup_strerror(error));
		goto free_cgroup;
	}

	ret = cgroup_init();
	if (ret) {
		fprintf(stderr, "%s: libcgroup initialization failed: %s\n",
			argv[0], cgroup_strerror(ret));
		goto free_cgroup;
	}

	for (i = 0; i < cfg_files.count; i++) {
		ret = cgroup_config_load_config(cfg_files.items[i]);
		if (ret) {
			fprintf(stderr, "%s; error loading %s: %s\n", argv[0],
					cfg_files.items[i],
					cgroup_strerror(ret));
			if (!error)
				error = ret;
		}
	}

free_cgroup:
	cgroup_free(&default_group);
err:
	cgroup_string_list_free(&cfg_files);
	return error;
}
